File: Algorithms/validateBinarySearchTree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        return validBST(root, null, null);
    }
    boolean validBST(TreeNode root, Integer min, Integer max) {
        if (root == null)
            return true;
        if (min != null && min >= root.val)
            return false;
        if (max != null && max <= root.val)
            return false;
        return validBST(root.left, min, root.val) 
            && validBST(root.right, root.val, max);
    }
}

File: Algorithms/validateBinarySearchTree.cpp
#include "leetcode.h"
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        return validBST(root, false, 0, false, 0);
    }
    bool validBST(TreeNode *root, bool minBound, int min, bool maxBound, int max) {
        if (root == NULL)
            return true;
        if (minBound && min >= root->val)
            return false;
        if (maxBound && max <= root->val)
            return false;
        return validBST(root->left, minBound, min, true, root->val) 
            && validBST(root->right, true, root->val, maxBound, max);
    }
};
